### PR TITLE
Introduce errors on incorrect target

### PR DESCRIPTION
**Pull request by wldeh**
Fail if a legacy JS target/compilation is registered, Kotlin/JS doesn't target Node.js, or Kotlin/WASM does not target D8. Change JsBenchmarkTarget.compilation type to KotlinJsIrCompilation

### CODE REVIEW

File: plugin/main/src/kotlinx/benchmark/gradle/BenchmarksExtension.kt [hunk 1]
```diff
@@ -63,12 +63,13 @@ open class BenchmarksExtension(val project: Project) {
...
                         is KotlinJsCompilation -> {
                             if (compilation.target.platformType != KotlinPlatformType.wasm) {
+                                check(compilation is KotlinJsIrCompilation) // Ensure compilation is KotlinJsIrCompilation
                                 JsBenchmarkTarget(this, name, compilation)
                             } else {
                                 check(compilation is KotlinJsIrCompilation)
                                 WasmBenchmarkTarget(this, name, compilation)
                             }
-                        }
+                        }
```
Comment (qurbonzoda) [inline plugin/main/src/kotlinx/benchmark/gradle/BenchmarksExtension.kt:72]: Please revert this change

File: plugin/main/src/kotlinx/benchmark/gradle/JsEngineExecTasks.kt [hunk 2]
```diff
@@ -28,7 +28,7 @@ fun Project.createJsEngineBenchmarkExecTask(
     }
 
     if (compilationTarget is KotlinWasmSubTargetContainerDsl) {
-        check(compilation is KotlinJsIrCompilation) { "Legacy Kotlin/JS is does not supported by D8 engine" }
+        check(compilation is KotlinJsIrCompilation) { "Legacy Kotlin/JS is not supported by D8 engine" }
```
Comment (qurbonzoda) [inline plugin/main/src/kotlinx/benchmark/gradle/JsEngineExecTasks.kt:31]: Is it possible to change the `compilation` argument type to `KotlinJsIrCompilation`?

File: integration/src/main/kotlin/kotlinx/benchmark/integration/ProjectBuilder.kt [hunk 3]
```diff
@@ -23,7 +29,10 @@ benchmark {
         ${targets.flatMap { it.value.lines(it.key) }.joinToString("\n        ")}
     }
 }
-            """.trimIndent()
+kotlin {
```
Comment (qurbonzoda) [inline integration/src/main/kotlin/kotlinx/benchmark/integration/ProjectBuilder.kt:32]: WDYT about moving up the `kotlin {}` section? As we documented in the README, the step for adding Kotlin targets precedes the steps for configuring benchmarks.

File: integration/src/main/kotlin/kotlinx/benchmark/integration/ProjectBuilder.kt [hunk 4]
```diff
@@ -23,7 +29,10 @@ benchmark {
...
+    ${kotlinConfig.lines().joinToString("\n    ")}
+}
+        """.trimIndent()
```
Comment (qurbonzoda) [inline integration/src/main/kotlin/kotlinx/benchmark/integration/ProjectBuilder.kt:35]: Minor: It's a nice practice to write opening and closing triple quotes with the same indentation.

File: integration/src/test/kotlin/kotlinx/benchmark/integration/InvalidTargetingTest.kt [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        val runner = project("kotlin-multiplatform") {
+            configuration("invalidJS") {
+                iterations = 1
+                iterationTime = 100
+                iterationTimeUnit = "ms"
+                reportFormat = "json"
+            }
+            register("jsTest")
+            kotlin {
+                js("jsTest", KotlinConfiguration.IR) {
```
Comment (qurbonzoda) [inline integration/src/test/kotlin/kotlinx/benchmark/integration/InvalidTargetingTest.kt:40]: Could you please also add a test for js with the legacy compiler?

File: integration/src/test/kotlin/kotlinx/benchmark/integration/InvalidTargetingTest.kt [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+            configuration("invalidWasm") {
+                iterations = 1
+                iterationTime = 100
+                iterationTimeUnit = "ms"
+                reportFormat = "json"
+            }
+            register("wasm")
+            kotlin {
+                wasm("wasm") {
+                    nodejs()
```
Comment (qurbonzoda) [inline integration/src/test/kotlin/kotlinx/benchmark/integration/InvalidTargetingTest.kt:19]: Could you please also add a test when the `browser()` environment is selected?

File: integration/src/test/kotlin/kotlinx/benchmark/integration/InvalidTargetingTest.kt [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+            configuration("invalidJS") {
+                iterations = 1
+                iterationTime = 100
+                iterationTimeUnit = "ms"
+                reportFormat = "json"
+            }
+            register("jsTest")
+            kotlin {
+                js("jsTest", KotlinConfiguration.IR) {
+                    d8()
```
Comment (qurbonzoda) [inline integration/src/test/kotlin/kotlinx/benchmark/integration/InvalidTargetingTest.kt:41]: Could you please also test when the `browser()` environment is selected?

**Review by qurbonzoda — CHANGES_REQUESTED** [review]
Note that 
* for `js` only `IR` compiler and `nodejs()` environment are supported.
* for `wasm` only `d8()`  environment is supported. 

Other configurations should be rejected with a clear failure message.

File: integration/src/test/kotlin/kotlinx/benchmark/integration/InvalidTargetingTest.kt [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        val runner = project("kotlin-multiplatform") {
+            configuration("invalidWasm") {
+                iterations = 1
+                iterationTime = 100
+                iterationTimeUnit = "ms"
+                reportFormat = "json"
+            }
+            register("wasm")
+            kotlin {
+                wasm("wasm") {
```
Comment (qurbonzoda) [inline integration/src/test/kotlin/kotlinx/benchmark/integration/InvalidTargetingTest.kt:18]: Recently, we decided to use "wasmJs" name for wasm target. See https://github.com/Kotlin/kotlinx-benchmark/pull/129

File: plugin/main/src/kotlinx/benchmark/gradle/BenchmarksExtension.kt (hunk 1 above)
Comment (wldeh) [reply plugin/main/src/kotlinx/benchmark/gradle/BenchmarksExtension.kt:72]: Thank you for catching this.

File: integration/src/main/kotlin/kotlinx/benchmark/integration/ProjectBuilder.kt (hunk 3 above)
Comment (wldeh) [reply integration/src/main/kotlin/kotlinx/benchmark/integration/ProjectBuilder.kt:32]: It is definitely more appropriate. Moved in the latest commit

File: integration/src/test/kotlin/kotlinx/benchmark/integration/InvalidTargetingTest.kt (hunk 8 above)
Comment (wldeh) [reply integration/src/test/kotlin/kotlinx/benchmark/integration/InvalidTargetingTest.kt:18]: Noted.

File: integration/src/main/kotlin/kotlinx/benchmark/integration/ProjectBuilder.kt (hunk 4 above)
Comment (wldeh) [reply integration/src/main/kotlin/kotlinx/benchmark/integration/ProjectBuilder.kt:35]: Thank you for catching this.

File: integration/src/test/kotlin/kotlinx/benchmark/integration/InvalidTargetingTest.kt (hunk 7 above)
Comment (wldeh) [reply integration/src/test/kotlin/kotlinx/benchmark/integration/InvalidTargetingTest.kt:41]: Done

File: integration/src/test/kotlin/kotlinx/benchmark/integration/InvalidTargetingTest.kt (hunk 6 above)
Comment (wldeh) [reply integration/src/test/kotlin/kotlinx/benchmark/integration/InvalidTargetingTest.kt:19]: Resolved

File: integration/src/test/kotlin/kotlinx/benchmark/integration/InvalidTargetingTest.kt (hunk 5 above)
Comment (wldeh) [reply integration/src/test/kotlin/kotlinx/benchmark/integration/InvalidTargetingTest.kt:40]: Resolved https://github.com/Kotlin/kotlinx-benchmark/pull/142/commits/d75202f29dce77121c4b1c6bcd829ff2e85da3af

**Review by qurbonzoda — CHANGES_REQUESTED** [review]

File: plugin/main/src/kotlinx/benchmark/gradle/JsSourceGeneratorTask.kt [hunk 9]
```diff
@@ -62,27 +62,11 @@ open class JsSourceGeneratorTask
 
     private fun load(lib: File): List<ModuleDescriptor> {
         val storageManager = LockBasedStorageManager("Inspect")
-        return if (ir) {
```
[the remark this answers is not in the document]
Comment (wldeh) [reply plugin/main/src/kotlinx/benchmark/gradle/JsSourceGeneratorTask.kt:65]: Yes. Removed in https://github.com/Kotlin/kotlinx-benchmark/pull/142/commits/94594d2f36bfaf8cfd2d2d2e713a68e965e0a9ef Thank you for pointing this out.

File: plugin/main/src/kotlinx/benchmark/gradle/JsSourceGeneratorTask.kt [hunk 10]
```diff
@@ -62,27 +62,11 @@ open class JsSourceGeneratorTask
...
+        // skip processing of empty dirs (fail if not to do it)
         if (lib.listFiles() == null) return emptyList()
         val dependencies = inputDependencies.files.filterNot { it.extension == "js" }.toSet()
         val module = KlibResolver.JS.createModuleDescriptor(lib, dependencies, storageManager)
         return listOf(module)
     }
-
-    private fun loadLegacy(lib: File, storageManager: StorageManager): List<ModuleDescriptor> {
-        val dependencies = inputDependencies.flatMap {
-            loadJsDescriptors(it, storageManager)
```
[the remark this answers is not in the document]
Comment (wldeh) [reply plugin/main/src/kotlinx/benchmark/gradle/JsSourceGeneratorTask.kt:82]: It is not. Removed in https://github.com/Kotlin/kotlinx-benchmark/pull/142/commits/58334d1b73c9f6a2e04f9fde1a12ba114b593a87

File: plugin/main/src/kotlinx/benchmark/gradle/JsEngineExecTasks.kt (hunk 2 above)
Comment (wldeh) [reply plugin/main/src/kotlinx/benchmark/gradle/JsEngineExecTasks.kt]: The type was not modified in this case but when it is an IR compilation then the type is KotlinJsIrCompilation.

File: plugin/main/src/kotlinx/benchmark/gradle/BenchmarksExtension.kt [hunk 11]
```diff
@@ -97,4 +97,4 @@ open class BenchmarksExtension(val project: Project) {
             }
         }
     }
-}
+}
```
Comment (qurbonzoda) [inline plugin/main/src/kotlinx/benchmark/gradle/BenchmarksExtension.kt:100]: Please try to avoid changes like this. It adds a needless entry to the file history.

File: plugin/main/src/kotlinx/benchmark/gradle/JsMultiplatformTasks.kt [hunk 12]
```diff
@@ -73,11 +73,10 @@ private fun Project.createJsBenchmarkGenerateSourceTask(
...
         title = target.name
-        ir = target.compilation is KotlinJsIrCompilation
         useBenchmarkJs = target.jsBenchmarksExecutor == JsBenchmarksExecutor.BenchmarkJs
         inputClassesDirs = compilationOutput.output.allOutputs
         inputDependencies = compilationOutput.compileDependencyFiles
         outputResourcesDir = file("$benchmarkBuildDir/resources")
         outputSourcesDir = file("$benchmarkBuildDir/sources")
     }
-}
+}
```
Comment (qurbonzoda) [inline plugin/main/src/kotlinx/benchmark/gradle/JsMultiplatformTasks.kt:82]: Please try to avoid unnecessary changes like this.

**Comment by qurbonzoda** [comment]
Merged manually

File: plugin/main/src/kotlinx/benchmark/gradle/BenchmarksExtension.kt (hunk 11 above)
Comment (wldeh) [reply plugin/main/src/kotlinx/benchmark/gradle/BenchmarksExtension.kt:100]: Noted. I am curious as to what the root cause of this showing up in the file history is given their is no change. Will be wary moving forward.

File: plugin/main/src/kotlinx/benchmark/gradle/JsMultiplatformTasks.kt (hunk 12 above)
Comment (wldeh) [reply plugin/main/src/kotlinx/benchmark/gradle/JsMultiplatformTasks.kt:82]: Noted.
